caledonia/src/bounded.rs

```diff
@@ -0,0 +1,451 @@
+    fn h2(setup: &Setup, r: &Round) -> bool {
+        let mut data = Vec::new();
+        data.push(r.v.to_ne_bytes().to_vec());
+        data.push(r.t.to_ne_bytes().to_vec());
+        for s in &r.s_list {
+            data.push(s.clone().to_vec());
+        }
+        let digest = utils::combine_hashes(data);
+        return utils::oracle(&digest, setup.q) == 0;
+    }
+
+    /// Depth-first search which goes through all potential round candidates
+    /// and returns first round candidate Round{t, x_1, ..., x_u)} such that:
+    /// - for all i ∈ [0, u-1], H0(x_i+1) ∈ bins[H1(t, x_1, ..., x_i)]
+    /// - H2(t, x_0, ..., x_u) = true
+    fn dfs(
+        setup: &Setup,
+        bins: &Vec<Vec<[u8; 32]>>,
+        round: &Round,
+        limit: &mut usize,
+    ) -> Option<Proof> {
+        // If the round candidate is at the last iteration (len(round) == u)
+        if round.s_list.len() == setup.u {
+            // if H2(t, x_0, ..., x_u) = true, return candidate as proof
+            if Proof::h2(setup, round) {
+                let r = round.v;
+                let d = round.t;
+                let items = round.s_list.clone();
+                return Some(Proof { r, d, items });
+            } else {
+                return None;
+            }
+        }
+        // Otherwise, update round candidate with all s in bins[H1(t, x_1, ..., x_i)] and continue
+        let result = bins[round.h_usize].iter().find_map(|&s| {
+            if *limit == 0 {
+                return None;
+            }
+            *limit -= 1;
+            Self::dfs(setup, bins, &Round::update(round, s), limit)
+        });
+        return result;
+    }
+
+    /// Alba's proving algorithm, based on a depth-first search algorithm.
+    /// Returns an empty proof if no suitable candidate is found.
+    fn prove_index(setup: &Setup, set: &Vec<[u8; 32]>, v: usize) -> (usize, Option<Proof>) {
+        // Initialising our n_p bins with all s ∈ Sp
+        let mut bins: Vec<Vec<[u8; 32]>> = Vec::new();
+        for _ in 1..(setup.n_p + 1) {
+            bins.push(Vec::new());
+        }
+        for &s in set.iter() {
+            // TODO: add H1(s),
+            // let (hs, index_s) = Proof::h0(setup, s);
+            bins[Proof::h0(setup, v, s)].push(s);
+        }
+        // Initialising max number of steps
+        let mut limit = setup.b;
+        // Attempting to generate a proof for $d$ rounds
+        for t in 1..(setup.d + 1) {
+            if limit == 0 {
+                return (0, None);
+            }
+            limit = limit - 1;
+            let round = Round::new(v, t, setup.n_p);
+            let res = Proof::dfs(setup, &bins, &round, &mut limit);
+            if res.is_some() {
+                return (limit, res);
+            }
+        }
+        // If no proof found, return None
+        return (limit, None);
+    }
+
+    /// Alba's proving algorithm, based on a depth-first search algorithm.
```

Comment:
This comment is the same as the previous function, which might be confusing for the prospective user approaching the library for the first time